stores/slabbuffer.go

```diff
@@ -0,0 +1,505 @@
+			if err := sqlStore.db.Delete(&buffer).Error; err != nil {
+				return nil, fmt.Errorf("failed to delete buffer %v: %v", buffer.ID, err)
+			}
+			if err := os.RemoveAll(filepath.Join(partialSlabDir, buffer.Filename)); err != nil {
+				return nil, fmt.Errorf("failed to remove buffer file %v: %v", buffer.Filename, err)
+			}
+			continue
+		}
+		// Open the file.
+		file, err := os.OpenFile(filepath.Join(partialSlabDir, buffer.Filename), os.O_RDWR, 0600)
+		if err != nil {
+			return nil, err
+		}
+		// Create the slab buffer.
+		var ec object.EncryptionKey
+		if err := ec.UnmarshalText(buffer.DBSlab.Key); err != nil {
+			return nil, err
+		}
+		sb := &SlabBuffer{
+			dbID:     buffer.ID,
+			filename: buffer.Filename,
+			slabKey:  ec,
+			maxSize:  int64(bufferedSlabSize(buffer.DBSlab.MinShards)),
+			file:     file,
+			dbSize:   buffer.Size,
+			size:     buffer.Size,
+		}
+		// Add the buffer to the manager.
+		gid := bufferGID(buffer.DBSlab.MinShards, buffer.DBSlab.TotalShards, uint32(buffer.DBSlab.DBContractSetID))
+		if buffer.Complete {
+			mgr.completeBuffers[gid] = append(mgr.completeBuffers[gid], sb)
+		} else {
+			mgr.incompleteBuffers[gid] = append(mgr.completeBuffers[gid], sb)
+		}
+		mgr.buffersByKey[sb.slabKey.String()] = sb
+	}
+	return mgr, nil
+}
+
+func bufferGID(minShards, totalShards uint8, contractSet uint32) bufferGroupID {
+	var bgid bufferGroupID
+	bgid[0] = minShards
+	bgid[1] = totalShards
+	binary.LittleEndian.PutUint32(bgid[2:], contractSet)
+	return bgid
+}
+
+func (mgr *SlabBufferManager) Close() error {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+	for _, buffers := range mgr.buffersByKey {
+		if err := buffers.file.Close(); err != nil {
+			return err
+		}
+	}
+	mgr.buffersByKey = nil
+	mgr.incompleteBuffers = nil
+	mgr.completeBuffers = nil
+	return nil
+}
+
+func (mgr *SlabBufferManager) AddPartialSlab(ctx context.Context, data []byte, minShards, totalShards uint8, contractSet uint) ([]object.PartialSlab, error) {
+	gid := bufferGID(minShards, totalShards, uint32(contractSet))
+
+	// Sanity check input.
+	slabSize := bufferedSlabSize(minShards)
+	if minShards == 0 || totalShards == 0 || minShards > totalShards {
+		return nil, fmt.Errorf("invalid shard configuration: minShards=%v, totalShards=%v", minShards, totalShards)
+	} else if contractSet == 0 {
+		return nil, fmt.Errorf("contract set must be set")
+	} else if len(data) > slabSize {
+		return nil, fmt.Errorf("data size %v exceeds size of a slab %v", len(data), slabSize)
+	}
+
+	// Deep copy available buffers. We don't want to block the manager while we
+	// perform disk I/O.
+	mgr.mu.Lock()
+	buffers := append([]*SlabBuffer{}, mgr.incompleteBuffers[gid]...)
+	mgr.mu.Unlock()
+
+	// Find a buffer to use. We use at most 1 existing buffer + 1 new buffer to
+	// avoid splitting the data over too many slabs.
+	var slab object.PartialSlab
+	var slabs []object.PartialSlab
+	var err error
+	var usedBuffers []*SlabBuffer
+	for _, buffer := range buffers {
```

Comment:
this could be replaced by a method on the manager that returns a locked buffer... not sure if that would be cleaner but it's worth considering? The deep copy + for loop and break is odd at first sight

Reply:
Imo this is fine. We have this pattern in a few places, especially maps. Pulling it out doesn't change the pattern.